### ISSUE TITLE
Clarify uninitialized values with examples

### ISSUE THREAD

**Issue by MollyDream** [p4lang/p4-spec#988]
The spec talks about uninitialized values in [section 8.22](https://p4.org/p4-spec/docs/P4-16-v1.2.2.html#sec-uninitialized-values-and-writing-invalid-headers), but given the amount of information, this section is very hard to follow and easy to misinterpret. I think adding more examples will clarify "a consistent value [does not have to be] returned across multiple such [reads]" of any uninitialized value, and will clarify what a "read" actually means.
Example 1:
```
bit<8> x;
bit<8> y = x+1;
bool r = (y==y);
```
The “read” of the uninitialized variable `x` occurs inside the expression evaluation `x+1`, taking an unspecified 8-bit value for the subexpression `x`. The value of `x+1` initializes `y` deterministically, so the following two reads of `y` gives the same value, leading to `r = true`.
Example 2:
```
bit<8> x;
bool r = (x==x);
```
The “read” of the uninitialized variable `x` occurs twice inside the expression evaluation `x == x`, each possibly resulting in a different value, so`r` can be either `true` or `false`.
Example 3:
```
void f(in bit<8> x, out bool y) {
   y = (x==x);
}
void g(...) {
  bit<8> a; 
  bool b;
  f(a,b);
}
```
Copy-in of an uninitialized variable to an `in` parameter of a function counts as a "read", resulting in a deterministic unspecified value of `x` in the function `f`. Therefore, `y = true`.
Example 4:
```
void f(out bit<8> x) {
  bool r1 = (x == x);
}
void g(...) {
  bit<8> a = 1; 
  f(a);
  bool r2 = (a==a);
}
```
Since `x` is an uninitialized `out` parameter inside the function `f`, `r1` can be either `true` or `false`. After executing `f`, copy-out of uninitialized variables **does not*** count as a read, so `a` becomes uninitialized, and `r2` can be either `true` or `false`.

\* This is based on our understanding of the spec. We think this design choice of "copy-in is a read but copy-out not" makes sense since it makes the function inlining smooth. Specifically, when the compiler inlines a function call, all the `in` parameters just have to be replaced by the arguments, and all the `out` arguments be re-declared; this pertains the behavior of uninitialized parameters in function inlining.

**Comment by jafingerhut** [comment]
@MollyDream I am all for examples clarifying some of these subtleties, so thanks for your write-up.

I agree that when I wrote the original text of this, I was not explicitly addressing what constitutes a "read" of a variable, and copy-in and copy-out were not explicitly addressed, and it is worth considering whether either or both of those are a "read" or not.

I was curious if you found anything in the current spec that would cause different behavior relative to uninitialized values for copy-in vs. copy-out?  Or perhaps your choice above is based upon consideration of the kinds of transformations that a correct P4 compiler can make to the original program?

**Comment by QinshiWang** [comment]
In general, I don't strongly support this, but just for information. The reason in favor of treating copy-out (and maybe also copy-in sometimes) as non-read is that it makes function inlining more efficient. For example,
```
void f(out bit<8> x, out bit<8> y) {
  y = 0;
}

{
  bit<8> x, y; // x is uninitilized
  f(x, y);
}
```
If copy-out is not read, we can safely transform it into
```
{
  bit<8> x, y; // x is uninitilized
  y = 0;
}
```
If copy-out is read, it's not safe to do it because
```
{
  bit<8> x, y; // x is uninitilized
  y = 0; // x is still uninitilized
}
```

**Comment by jafingerhut** [comment]
During 2022-Jan-03 LDWG meeting, Mihai & Andy came to the realization that perhaps p4c's LocalCopyPropagation pass can result in output programs that are more nondeterministic than the input program, unless the compiler can somehow restrict the use of this transformation to cases where the variables/fields being accessed in the expression are "at least as deterministic as" the original program, e.g. all of them have initialized values.

AI Andy: Try to find a small P4_16 program where the current p4c implementation's LocalCopyPropagation pass produces a program that is strictly more non-deterministic than the input program, and add it here.

In general, I am fine with any definition of the P4 language spec here for which we know which compiler transformations are correct (i.e. result is at least as determinstic as the input program), and which are incorrect, and the incorrect ones are optimization passes we are willing to live without.

**Comment by jfingerh** [comment]
For anyone curious, there is a very short P4 program for which today's p4c can increase the number of possible behaviors of an input P4 program, because of the LocalCopyPropagation pass, when at least some of the variables in the resulting expressions are uninitialized: https://github.com/jafingerhut/p4-guide/tree/master/nondeterminism#demonstration-of-localcopypropagation-increasing-nondeterminism-of-input-program

**Comment by mbudiu-vmw** [comment]
Here is an article about uninitialized values from someone who wrote a Ph.D. thesis on semantics: https://www.ralfj.de/blog/2019/07/14/uninit.html

**Comment by jfingerh** [comment]
Thanks for that link.  It looks highly relevant to P4's current definition of the behavior of uninitialized variables, from what I can tell.  I haven't read all the linked discussions yet, as I only spent a couple of minutes on the directly linked article.  From that alone, I get the impression that the author has thought about these issues, but I do not yet know if they have recommendations that help us answer the question posed by this issue.

**Comment by mbudiu-vmw** [comment]
The main point of his text is that *we* (the LDWG) should define what "undefined" means, and not specific implementations, like Tofino or BMv2. The abstract machine for the language definition, aka the spec, has choices to make, and we make them. We can choose whether a read of an undefined variable "freezes" the value or does not. Both are legitimate choices.

**Comment by jnfoster** [comment]
Ralf was almost a P4 h4x0r! (I was this close to hiring him as a postdoc.) Sadly, he decided to go to some place called MIT.

**Comment by jfingerh** [comment]
"The main point of his text is that we (the LDWG) should define what "undefined" means, and not specific implementations, like Tofino or BMv2."  I agree with this statement, and believed it before I read the linked article.  I still do not know the best way to answer the questions raised by this issue, though :-(

**Comment by mbudiu-vmw** [comment]
There may be no "best" answer. This is why language design is hard. As you see, other communities struggle with the same problems.

**Comment by QinshiWang** [comment]
Just recall the reason that we need partially initialized value by bit. Consider we assigning `x` by parts:
```
void f(in bit<4> y, in bit<4> z) {
   bit<8> x;
   x[3:0] = y;
   x[7:4] = z;
   ...
}
The only intuitive interpretation of `x` after two partial assignments is `x = y ++ z`. We must have bitwise value to support it.

**Comment by jnfoster** [comment]
In the interest of tidying up the set of active issues on the P4 specification repository, I'm marking this as "stalled" and closing it. Of course, we can always re-open it in the future if there is interest in resurrecting it.
